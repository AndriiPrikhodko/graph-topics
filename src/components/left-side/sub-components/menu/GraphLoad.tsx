import React, { useState, useEffect, MouseEvent, ChangeEvent } from 'react'
import { setGraphData } from '../../../../reducers/graph.reducer'
import { graphAPI } from '../../../../api/graph.api'
import {selfDataAdapter} from '../../../../helpers/data-adapter/self-data-adapter'
import Dropdown from './Dropdown'
import { useSelector, useDispatch } from 'react-redux'
import { loadGraphList } from '../../../../actions/graph-api.action'
import Button from '../../../shared/button'
import { setGraphDataLocal } from '../../../../helpers/local-storage'


const GraphLoad: React.FC = () => {
    const dispatch = useDispatch()
    const graphList = useSelector((store: Store) => store.graphList)
    const [menuOption, setMenuOption] = useState<string>('')
    const [menuValue, setMenuValue] = useState<string>('')

    useEffect(() => {
        try {
            dispatch(loadGraphList())
        } catch (err) {
            console.error(err)
        }
    }, [dispatch])


    const handleMenuItemClick = (menuItemProps: MouseEvent & {defaultValue?: string}) => {
        if (menuItemProps.defaultValue) setMenuOption(menuItemProps.defaultValue)
    }

    const handleMenuChange = () => {
        setMenuValue(menuOption)
    }

    const handleLoadGraphClick = () => {
        if (menuOption) {
            const args = {
                graphName: menuOption
            }

            graphAPI.getGraph(args)
                .then(res => selfDataAdapter(res))
                .then(graphData => {
                    setGraphDataLocal(graphData)
                    dispatch(setGraphData(graphData))
                })
        }
    }

    return <div className="graph-menu" data-testid='load-graph-list'>
        <Dropdown
            trigger={<input
                className="dropdown-trigger"
                value={menuOption}
                onChange={handleMenuChange}
                data-testid='load-dropdown'
            >
            </input>}
            menu={graphList.data.map((graphName: string) => <input
                className='dropdown-input'
                defaultValue={graphName}
                onClick={handleMenuItemClick}
                readOnly></input>)
            }
        />
        <Button
            label='Load'
            onClick={handleLoadGraphClick}
            testid='load-graph-button'
            className='menu-button-left'
        />
    </div>
}

export default GraphLoad