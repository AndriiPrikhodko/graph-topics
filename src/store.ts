import { configureStore } from '@reduxjs/toolkit'
import rootReducer from './reducers'
import thunk from 'redux-thunk'

const store = configureStore({
  reducer: {
    graphData: rootReducer.graphSlice,
    graphList: rootReducer.graphListSlice
  },
  middleware: [thunk],
})

export default store;