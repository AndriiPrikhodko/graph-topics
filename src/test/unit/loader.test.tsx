import {  test } from 'vitest'
import { render } from '@testing-library/react'
import GraphLoad from '../../components/left-side/sub-components/menu/GraphLoad'
import store from '../../store'
import { Provider } from 'react-redux'

test('loader', () => {
    const { getByTestId } = render(
        <Provider store={store}>
            <GraphLoad />
        </Provider>
    )
    const loaderEl = getByTestId('load-graph-list')
    expect(loaderEl).toMatchSnapshot()
})

