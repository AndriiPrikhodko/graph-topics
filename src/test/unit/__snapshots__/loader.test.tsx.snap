// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loader 1`] = `
<div
  class="graph-menu"
  data-testid="load-graph-list"
>
  <div
    class="dropdown"
  >
    <input
      class="dropdown-trigger"
      data-testid="load-dropdown"
      value=""
    />
  </div>
  <button
    class="standard-button menu-button-left"
    data-testid="load-graph-button"
  >
    Load
  </button>
</div>
`;
